synapse_token_authenticator/claims_validator.py

```diff
@@ -0,0 +1,179 @@
+from dataclasses import dataclass
+from typing import List, Optional, Any, TypeAlias, Union
+from synapse_token_authenticator.utils import get_path_in_dict
+import re
+
+Validator: TypeAlias = Union[
```

Comment:
I agree, we could use something existing

Existing solution:
1. No need to maintain our own
2. More out of the box features
3. Validation errors reporting, for debuging

Current solution:
1. Flexibility: having dict-style syntax for real explicitness + shorter list-style syntax for density (one simple check -- one line)
2. **D**SL: being domain specific, it really focuses on validating jwt claims. This means we don't have anything extra, like comparing numbers (which could be useful actually), the main terminal checkers are `equal` and `regex` (not sure we need regex though, but could be useful)
3. Is fairly simple, so there's not much maintanence cost. If we ever need something significantly more complicated (rather than adding couple more terminal validators), we better switch to other engine/dsl

The bigges pros to current solution is 1 and 3 IMO, simplicity also implies it's easier to learn.

Just for the context, there's few stages of validation:
1. validate structure/schema/shape (e.g. [RFC 8927 - JSON Type Definition](https://datatracker.ietf.org/doc/html/rfc8927))
2. validate data inside (e.g. a list should be non empty, a field should > 3, etc.)
3. validate data relative to itself (e.g. field `a` should be > than field `b`)
4. validate data with external static context (basically the same as 2 and 3 but depending on config)
5. validate data with external resource (e.g. if `data.is_admin` then fetch `/some_url`  and assert `data.a` > `some_url_result.a`

So validation is interesting topic and gets harder to implement configurable at each level. For our purpose we don't need to go above 2 though

Reply:
Mind leaving parts of this comment in the code for future reference? It'd be easy for a future team to not realize they can just try to switch this out for a generic solution instead of adding more features.

Reply:
Added: [synapse_token_authenticator/claims_validator.py:1](https://github.com/famedly/synapse-token-authenticator/blob/92b1ff2ee44ef3bbada23b85a89d5a3f542dcba6/synapse_token_authenticator/claims_validator.py#L1-L17)